[SPARK-22880][SQL] Add cascadeTruncate option to JDBC datasource

danielvdende opened the pull request:
This commit adds the `cascadeTruncate` option to the JDBC datasource
API, for databases that support this functionality (PostgreSQL and
Oracle at the moment). This allows for applying a cascading truncate
that affects tables that have foreign key constraints on the table
being truncated.

## What changes were proposed in this pull request?

Add `cascadeTruncate` option to JDBC datasource API. Allow this to affect the 
`TRUNCATE` query for databases that support this option.

## How was this patch tested?
Existing tests for `truncateQuery` were updated. Also, an additional test was added
to ensure that the correct syntax was applied, and that enabling the config for databases
that do not support this option does not result in invalid queries.

danielvdende committed:
[SPARK-22880][SQL] Add cascadeTruncate option to JDBC datasource

This commit adds the cascadeTruncate option to the JDBC datasource
API, for databases that support this functionality (PostgreSQL and
Oracle at the moment). This allows for applying a cascading truncate
that affects tables that have foreign key constraints on the table
being truncated.

danielvdende commented:
@dongjoon-hyun this is the functionality we discussed in PR for SPARK-22729, would be great to hear your opinion on this :)

AmplabJenkins commented:
Can one of the admins verify this patch?

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala:
```diff
@@ -120,11 +121,12 @@ abstract class JdbcDialect extends Serializable {
    */
   @Since("2.3.0")
-  def getTruncateQuery(table: String): String = {
+  def getTruncateQuery(table: String, cascade: Boolean = false): String = {
```
Could you remove the default value here?
It could be safe to prevent accidental inheritance of this wrong implementation.

danielvdende replied:
That would force all dialects to implement this method though, which would lead to unnecessary code duplication. Moreover, removing the default value here would change the public API, as it would force others who have written custom dialects to change calls to this method.

dongjoon-hyun replied:
That is the main reason I request the change. I don't think this implementation is correct on for all dialects. Anyway, it's only my opinion.

danielvdende replied:
I think the default for all dialects should be false, regardless of whether the cascade feature is even supported. And for those for which it is supported, it should default to false. It's a feature that you should only use if you explicitly want to cascade imho.

danielvdende commented:
I think the default for all dialects should be false, regardless of whether the cascade feature is even supported. And for those for which it is supported, it should default to false.

gatorsmile commented:
Thank you for your contribution! I am doubting the value of this feature.

If you are interested in the JDBC-related work, could you take https://issues.apache.org/jira/browse/SPARK-22731? 

Thanks!

danielvdende commented:
@gatorsmile could you explain why you have doubts about the feature? Thanks!

danielvdende commented:
@gatorsmile would be great to hear why you doubt the value of the feature :). I know that for us it would be extremely valuable (at the moment we have to do an extra step in our data pipeline because this feature is missing in Spark), but of course we're not the only ones using Spark.

gatorsmile commented on docs/sql-programming-guide.md:
```diff
@@ -1339,6 +1339,13 @@ the following case-insensitive options:
+  <tr>
+    <td><code>cascadeTruncate</code></td>
+    <td>
+        This is a JDBC writer related option. If enabled and supported by the JDBC database (PostgreSQL and Oracle at the moment), this options allows execution of a <code>TRUNCATE TABLE t CASCADE</code>. This will affect other tables, and thus should be used with case. This option applies only to writing.
```
This conf only takes an effect when `SaveMode.Overwrite` is enabled and `truncate` is set to true. My doubt is the usage scenario of this extra conf does not benefit most users. Could we hold this PR? When the other users have similar requests, we can revisit it?

danielvdende replied:
I think it raises the question of how complete/incomplete the Spark JDBC API should be, and what the use cases are that it should serve. For the most simple cases, in which no key constraints are set between tables, you won't need this option. However, as soon as foreign key constraints are introduced, it is very important. I agree that not every functionality from SQL (dialects) should be included, but I personally feel this is quite fundamental functionality. 

Moreover, as it's configuration option users that don't want it also don't have to use it. I think we also discussed this functionality in previous PR with @dongjoon-hyun  here: [SPARK-22729](https://github.com/apache/spark/pull/19911)

dongjoon-hyun replied:
Hi, @danielvdende .
I think you can ignore my comment in previous PR. There were many directional comments on that PR and it's not the final one. Your previous PR is merged by @gatorsmile .

For me, I also still don't agree on the default value inside `JdbcDialects.scala` in this PR.

danielvdende replied:
@dongjoon-hyun apologies if I misrepresented your comments/opinions from the previous PR, wasn't my intention :-).

I've given it some more thought, and I can see you point about the default value. I'll make the change we discussed.

danielvdende replied:
@dongjoon-hyun one thing I'm thinking of (just curious to hear your opinion): could we use the value of `isCascadingTruncateTable` for each dialect as the default value for the `cascade` boolean? In that way, there is only a single boolean per dialect specifying what the default behaviour is with regard to cascading during truncates.

Stephan202 replied:
About the documentation: `s/used with case/used with care/`

danielvdende committed:
Change cascade parameter to option

This allows us to use the isCascadingTruncateTable function as
default for each dialect

danielvdende commented:
I've made some changes to the code; the hardcoded `false` value for `cascade` in `JdbcDialects` is now replaced by a default value of `isCascadingTruncateTable`. This is also the case for each individual dialect, with a pattern match in the dialects for Postgres and Oracle, as they support the cascading behaviour. Curious to hear your thoughts on this @dongjoon-hyun @gatorsmile

danielvdende commented:
@dongjoon-hyun @gatorsmile any further thoughts on this?

danielvdende commented:
@dongjoon-hyun @gatorsmile any further update?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

Stephan202 commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/DB2Dialect.scala:
```diff
@@ -49,4 +49,16 @@ private object DB2Dialect extends JdbcDialect {
+  * The SQL query used to truncate a table.
+  * @param table The JDBCOptions.
+  * @param cascade Whether or not to cascade the truncation. Default value is the
+  *                value of isCascadingTruncateTable()
```
For dialects not supporting cascading, should the documentation perhaps indicate that the `cascade` parameter is ignored?

Stephan202 reviewed (commented):
We could use this feature 👍 .

danielvdende commented:
@Stephan202 thanks for pointing out those docs issues, just pushed the changes :-).
@gatorsmile @dongjoon-hyun would you have a chance to take a look at this again?

Fokko commented:
Any idea when this will be merged into master? We could use this since we are ditching sqoop 👍

danielvdende commented:
@dongjoon-hyun @gatorsmile sorry to keep asking, but could you let me know when we can get this merged?

Fokko commented:
We're in the process of integrating Spark in Airflow, and support for the `cascadeTruncate` is required to make this succeed. First steps are here: https://github.com/apache/incubator-airflow/pull/3021.

Would be great if we can get this merged asap so we can continue testing. Cheers

gatorsmile commented:
ok to test

SparkQA commented:
**[Test build #87484 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87484/testReport)** for PR 20057 at commit [`3a7dda4`](https://github.com/apache/spark/commit/3a7dda4a0df8ef684d8fb803a98434c170953f4c).

gatorsmile commented:
@danielvdende @Fokko We definitely want to help the community replace Sqoop by Spark SQL. However, `truncate` is only used when users use <code>SaveMode.Overwrite</code> to write the external JDBC tables. In this specific scenario, Spark will truncate an existing table instead of dropping and recreating it. 

Could you show me the key missing features that are available in Sqoop but not in Spark SQL JDBC connectors?

SparkQA commented:
**[Test build #87484 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87484/testReport)** for PR 20057 at commit [`3a7dda4`](https://github.com/apache/spark/commit/3a7dda4a0df8ef684d8fb803a98434c170953f4c).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/87484/
Test FAILed.

danielvdende commented:
Tests are failing on a spark streaming test. I think it's probably because of the age of this PR, will rebase to get the changes into the PR that were merged into master since I opened the PR

SparkQA commented:
**[Test build #87493 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87493/testReport)** for PR 20057 at commit [`6c0d3df`](https://github.com/apache/spark/commit/6c0d3dfd415e5630dbb02ce65c6adf3db419bdec).

SparkQA commented:
**[Test build #87493 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87493/testReport)** for PR 20057 at commit [`6c0d3df`](https://github.com/apache/spark/commit/6c0d3dfd415e5630dbb02ce65c6adf3db419bdec).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/87493/
Test FAILed.

Fokko commented:
Hi @gatorsmile, thanks for putting it to the tests. The main reason why I personally dislike Sqoop is:

- **Legacy.** The old map-reduce should be buried in the upcoming years. As a data engineering consultant, I see more people questioning the whole Hadoop stack. Using Sqoop you still need to run map-reduce tasks, and this isn't easy on other platforms like kubernetes.
- **Stability.** I see Sqoop jobs fail quite often, and there isn't a nice way of retrying this in an atomic way. For example, when having a Sqoop job on Airflow, we cannot simply retry the operation. We when we import data from a rmdbs to hdfs, we have to make sure that the target directory of the previous run has been deleted.

This is also where Spark-jdbc comes in, for example, in the future we would like to delete single partitions, but this is wip. Maybe @danielvdende can elaborate a bit on their use-case.

danielvdende commented:
Hmm, not it fails the OrcQuerySuite. This PR doesn't touch any of the Orc implementation in Spark. Could this be a flaky test @gatorsmile ?
```org.scalatest.exceptions.TestFailedDueToTimeoutException: The code passed to eventually never returned normally. Attempted 12 times over 10.158754687999998 seconds. Last failure message: There are 1 possibly leaked file streams..```

danielvdende commented:
Hi guys, @Fokko @gatorsmile, completely agree with what @Fokko mentioned, our main reason for wanting to get away from Sqoop is also for stability reasons and to get rid of MapReduce in preparation for our move to Kubernetes (or something similar). We've also seen it to be much faster than Sqoop. In terms of why we need the feature in this PR: we have some tables in PostgreSQL that have foreign keys linking them. We have also specified a schema for these tables. If we use the drop-and-recreate option, Spark will determine the schema, overriding our PostgreSQL schema. Obviously, these should match up, but I personally don't like that Spark can do this (and that you can't explicitly tell it not to). 

Because of this behaviour, we currently require 2 tasks in Airflow (as @Fokko mentioned) to ensure the tables are truncated, but the schema stays in place. This PR would enable us to specify in a single, idempotent (Airflow) task that we want to truncate the table before putting new data in there. The cascade enables us to not break foreign key relations and cause errors.

To be clear, this therefore isn't emulating a Sqoop feature (as a Sqoop task isn't idempotent), but is in fact improving on what Sqoop offers.

gatorsmile commented:
Our overwrite semantics is confusing to most. We need to correct it in the next release, i.e., Spark 2.4.

Even if we try our best to keep the schema of the original table, the actual CREATE TABLE statements still take many vendor-specific info. It is hard for us to rebuild all of them. I can understand your use case for truncate. 

I am sorry this will not be part of Spark 2.3 release. We will include it in the next release. You can still do the change in your forked Spark. 

Just feel free to let us know if you find anything that we should do in Spark SQL JCBC to match the corresponding ones in SQOOP. Thanks!

gatorsmile commented:
This test is a flaky test. Your changes did not fail any test case. I will review your PR after the 2.3 release. Thanks again!

cc @dongjoon-hyun Do you want to take a look at this?

dongjoon-hyun commented:
Thank you for pinging me, @gatorsmile . Yep. I'll take a look both `OrcQuerySuite` and this PR in this morning. Sorry for late response, @danielvdende .

danielvdende commented:
Thanks guys! @gatorsmile @dongjoon-hyun 
Happy to help out expanding Spark SQL JDBC where necessary to match and improve on sqoop :-)

dongjoon-hyun commented on docs/sql-programming-guide.md:
```diff
@@ -1372,6 +1372,13 @@ the following case-insensitive options:
+  <tr>
+    <td><code>cascadeTruncate</code></td>
+    <td>
+        This is a JDBC writer related option. If enabled and supported by the JDBC database (PostgreSQL and Oracle at the moment), this options allows execution of a <code>TRUNCATE TABLE t CASCADE</code>. This will affect other tables, and thus should be used with care. This option applies only to writing.
```
Maybe, do we need a default value description like `It defaults to <code>false</code>`?

danielvdende replied:
As mentioned in another comment, I think we should use the value of `isCascadingTruncateTable` as the default, rather than always `false`. Seems like the correct use of that variable. I can add a sentence to the docs specifying that.

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCOptions.scala:
```diff
@@ -119,6 +119,8 @@ class JDBCOptions(
   // if to truncate the table from the JDBC database
   val isTruncate = parameters.getOrElse(JDBC_TRUNCATE, "false").toBoolean
+
+  val isCascadeTruncate: Option[Boolean] = parameters.get(JDBC_CASCADE_TRUNCATE).map(_.toBoolean)
```
Sorry. I'm not sure that I lost the previous context, but what about `getOrElse` like `val isTruncate` on line 121?

danielvdende replied:
The reason I didn't do that is because of the existence of the `isCascadingTruncateTable` function for each dialect. According to the docs, that indicates whether or not a `TRUNCATE TABLE` command results in cascading behaviour by default for a given dialect. I thought it would be nice to then use that value as the default value for `isCascadeTruncate`. In that way, if there ever is a dialect that cascades truncations by default, we don't 'hardcode' a default value.

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala:
```diff
@@ -102,7 +102,12 @@ object JdbcUtils extends Logging {
+          options.isCascadeTruncate))
+      } else {
+        statement.executeUpdate(dialect.getTruncateQuery(options.table))
+      }
```
Simply, one line if we are using `boolean` instead of `Option`?
```
statement.executeUpdate(dialect.getTruncateQuery(options.table, options.isCascadeTruncate))
```

danielvdende replied:
see the previous comment on why this is an `Option[Boolean]`

dongjoon-hyun replied:
I see.

dongjoon-hyun replied:
What about the following?
```scala
val truncateQuery = if (options.isCascadeTruncate.isDefined) {
  dialect.getTruncateQuery(options.table, options.isCascadeTruncate))
} else {
  dialect.getTruncateQuery(options.table)
}
statement.executeUpdate(truncateQuery)
```

dongjoon-hyun replied:
Ping.

danielvdende replied:
Sorry, missed this comment. I'll make the change (@gatorsmile I think this is the comment you meant, right?)

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/AggregatedDialect.scala:
```diff
@@ -64,7 +64,15 @@ private class AggregatedDialect(dialects: List[JdbcDialect]) extends JdbcDialect
-  override def getTruncateQuery(table: String): String = {
-    dialects.head.getTruncateQuery(table)
+  /**
+  * The SQL query used to truncate a table.
```
Indentation?

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/AggregatedDialect.scala:
```diff
@@ -64,7 +64,15 @@ private class AggregatedDialect(dialects: List[JdbcDialect]) extends JdbcDialect
+  * @return The SQL query to use for truncating a table
+  */
+  override def getTruncateQuery(table: String,
+                                cascade: Option[Boolean] = isCascadingTruncateTable): String = {
```
Indentation?

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/DB2Dialect.scala:
```diff
@@ -49,4 +49,16 @@ private object DB2Dialect extends JdbcDialect {
   override def isCascadingTruncateTable(): Option[Boolean] = Some(false)
+
+  /**
+  * The SQL query used to truncate a table.
```
Indentation.

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/DB2Dialect.scala:
```diff
@@ -49,4 +49,16 @@ private object DB2Dialect extends JdbcDialect {
+  * @return The SQL query to use for truncating a table
+  */
+  override def getTruncateQuery(table: String,
+                                cascade: Option[Boolean] = isCascadingTruncateTable): String = {
```
ditto.

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/MsSqlServerDialect.scala:
```diff
@@ -42,4 +42,16 @@ private object MsSqlServerDialect extends JdbcDialect {
   override def isCascadingTruncateTable(): Option[Boolean] = Some(false)
+
+  /**
+  * The SQL query used to truncate a table.
```
ditto.

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/MsSqlServerDialect.scala:
```diff
@@ -42,4 +42,16 @@ private object MsSqlServerDialect extends JdbcDialect {
+  * @return The SQL query to use for truncating a table
+  */
+  override def getTruncateQuery(table: String,
+                                cascade: Option[Boolean] = isCascadingTruncateTable): String = {
```
ditto.

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/MySQLDialect.scala:
```diff
@@ -46,4 +46,16 @@ private case object MySQLDialect extends JdbcDialect {
   override def isCascadingTruncateTable(): Option[Boolean] = Some(false)
+
+  /**
+  * The SQL query used to truncate a table.
```
ditto.

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/MySQLDialect.scala:
```diff
@@ -46,4 +46,16 @@ private case object MySQLDialect extends JdbcDialect {
+  * @return The SQL query to use for truncating a table
+  */
+  override def getTruncateQuery(table: String,
+                                cascade: Option[Boolean] = isCascadingTruncateTable): String = {
```
ditto.

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/OracleDialect.scala:
```diff
@@ -94,5 +94,20 @@ private case object OracleDialect extends JdbcDialect {
   }
 
+  /**
+  * The SQL query used to truncate a table.
```
ditto.

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/OracleDialect.scala:
```diff
@@ -94,5 +94,20 @@ private case object OracleDialect extends JdbcDialect {
+  * @return The SQL query to use for truncating a table
+  */
+  override def getTruncateQuery(table: String,
+                                cascade: Option[Boolean] = isCascadingTruncateTable): String = {
```
ditto.

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala:
```diff
@@ -89,11 +89,17 @@ private object PostgresDialect extends JdbcDialect {
-  * @param table The name of the table.
+  * @param table The table to truncate
+  * @param cascade Whether or not to cascade the truncation. Default value is the value of
+    *                isCascadingTruncateTable()
```
ditto.

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/TeradataDialect.scala:
```diff
@@ -31,4 +31,16 @@ private case object TeradataDialect extends JdbcDialect {
   }
+
+  /**
+  * The SQL query used to truncate a table.
```
ditto.

dongjoon-hyun commented on sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala:
```diff
@@ -860,14 +860,41 @@ class JDBCSuite extends SparkFunSuite
+  test("SPARK-22880: Truncate table with CASCADE by jdbc dialect") {
+    // cascade in a truncate should only be applied for databases that support this,
+    // even if the parameter is passed.
+    val MySQL = JdbcDialects.get("jdbc:mysql://127.0.0.1/db")
```
Since this is variable, what about 
- `MySQL` -> `mysql`
- `Postgres` -> `postgres`

danielvdende replied:
Good point, I saw a few other tests in there (untouched by the PR initially) that had this, fixed those now too.

dongjoon-hyun commented on sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala:
```diff
@@ -860,14 +860,41 @@ class JDBCSuite extends SparkFunSuite
+    val db2 = JdbcDialects.get("jdbc:db2://127.0.0.1/db")
+    val h2 = JdbcDialects.get(url)
+    val derby = JdbcDialects.get("jdbc:derby:db")
+    val oracle = JdbcDialects.get("jdbc:oracle://127.0.0.1/db")
```
Could you add `teradata` test case, too?

danielvdende replied:
Done 👍

dongjoon-hyun commented:
Hi, @danielvdende. I finished my first round review.

dongjoon-hyun reviewed (changes requested):
I'll review again after updating. I understand it's important for Apache Airflow, but it's for Apache Spark 2.4 here.

danielvdende commented:
Thanks for your review @dongjoon-hyun 👍 I've corrected all the indentation problems (didn't show up locally when running scalastyle checks for some reason). Added comments where necessary for the other points. 

Sure, no problem to hold this until Spark 2.4, would be great if it could land there.

SparkQA commented:
**[Test build #87524 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87524/testReport)** for PR 20057 at commit [`3921368`](https://github.com/apache/spark/commit/39213688b9389e1a2aa4763a8a09c3d1ca402e6e).

SparkQA commented:
**[Test build #87524 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87524/testReport)** for PR 20057 at commit [`3921368`](https://github.com/apache/spark/commit/39213688b9389e1a2aa4763a8a09c3d1ca402e6e).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/87524/
Test FAILed.

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/DerbyDialect.scala:
```diff
@@ -41,4 +41,16 @@ private object DerbyDialect extends JdbcDialect {
   }
+
+  /**
+  * The SQL query used to truncate a table.
```
Indentation.

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/DerbyDialect.scala:
```diff
@@ -41,4 +41,16 @@ private object DerbyDialect extends JdbcDialect {
+  * @return The SQL query to use for truncating a table
+  */
+  override def getTruncateQuery(table: String,
+                                cascade: Option[Boolean] = isCascadingTruncateTable): String = {
```
indentation.

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala:
```diff
@@ -120,11 +121,13 @@ abstract class JdbcDialect extends Serializable {
   @Since("2.3.0")
-  def getTruncateQuery(table: String): String = {
+  def getTruncateQuery(table: String,
+                       cascade: Option[Boolean] = isCascadingTruncateTable): String = {
```
@danielvdende .
Could you take a look the guide? This guide is helpful in Apache Spark community.
- https://github.com/databricks/scala-style-guide

dongjoon-hyun replied:
For this one, we need to keep the original one for backward compatibility like the following?
```
  @Since("2.3.0")
  def getTruncateQuery(table: String): String = {
      getTruncateQuery(table, isCascadingTruncateTable)
  }

  @Since("2.4.0")
  def getTruncateQuery(table: String, cascade: Option[Boolean] = isCascadingTruncateTable)
```

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/DB2Dialect.scala:
```diff
@@ -49,4 +49,17 @@ private object DB2Dialect extends JdbcDialect {
+
+  /**
+   * The SQL query used to truncate a table.
+   * @param table The JDBCOptions.
```
`The JDBCOptions`?

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/MsSqlServerDialect.scala:
```diff
@@ -42,4 +42,17 @@ private object MsSqlServerDialect extends JdbcDialect {
+
+  /**
+   * The SQL query used to truncate a table.
+   * @param table The JDBCOptions.
```
`The JDBCOptions`?

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/MsSqlServerDialect.scala:
```diff
@@ -42,4 +42,17 @@ private object MsSqlServerDialect extends JdbcDialect {
+   * The SQL query used to truncate a table.
+   * @param table The JDBCOptions.
+   * @param cascade Whether or not to cascade the truncation. Default value is the
+   *                value of isCascadingTruncateTable(). Ignored for MsSql as it is unsupported.
```
I think we can ignore `Ignored for MsSql as it is unsupported.` here.
If you want, you can make a function comment of `override def isCascadingTruncateTable(): Option[Boolean] = Some(false)`.

dongjoon-hyun replied:
And, for all dialects, please remove it or move it consistently.

danielvdende replied:
Hmm, I disagree with this, the reason I added it was to show users that even if they specified `cascadeTruncate` being `true`, it doesn't take effect for some databases (like MsSQL). The `isCascadingTruncateTable` only shows the default behaviour, which for all databases at the moment is false.

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/MySQLDialect.scala:
```diff
@@ -46,4 +46,17 @@ private case object MySQLDialect extends JdbcDialect {
+
+  /**
+   * The SQL query used to truncate a table.
+   * @param table The JDBCOptions.
```
ditto.

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/OracleDialect.scala:
```diff
@@ -94,5 +94,21 @@ private case object OracleDialect extends JdbcDialect {
 
+  /**
+   * The SQL query used to truncate a table.
+   * @param table The JDBCOptions.
```
ditto.

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/OracleDialect.scala:
```diff
@@ -94,5 +94,21 @@ private case object OracleDialect extends JdbcDialect {
+   *                value of isCascadingTruncateTable()
+   * @return The SQL query to use for truncating a table
+   */
+  override def getTruncateQuery(
```
Could you move this function after `isCascadingTruncateTable` like the other files?

danielvdende replied:
Done

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/AggregatedDialect.scala:
```diff
@@ -64,7 +64,16 @@ private class AggregatedDialect(dialects: List[JdbcDialect]) extends JdbcDialect
-    dialects.head.getTruncateQuery(table)
+  /**
+   * The SQL query used to truncate a table.
+   * @param table The JDBCOptions.
```
?

dongjoon-hyun commented on docs/sql-programming-guide.md:
```diff
@@ -1372,6 +1372,13 @@ the following case-insensitive options:
+  <tr>
+    <td><code>cascadeTruncate</code></td>
+    <td>
+        This is a JDBC writer related option. If enabled and supported by the JDBC database (PostgreSQL and Oracle at the moment), this options allows execution of a <code>TRUNCATE TABLE t CASCADE</code>. This will affect other tables, and thus should be used with care. This option applies only to writing. It defaults to the default cascading truncate behaviour of the JDBC database in question, s ...
```
Ur, for PostgreSQL, we are generating `TRUNCATE TABLE ONLY $table CASCADE`.

danielvdende replied:
Yeah, so the "problem" there is that PostgreSQL has inheritance across tables (as we discovered in [SPARK-22729](https://github.com/apache/spark/pull/19911). To be completely transparent to users, I think the `ONLY` part of the query for PostgreSQL should also be configurable, but I think that's outside the scope of this PR. I've added a comment saying that for PostgreSQL a `ONLY` is added to prevent descendant tables from being truncated.

dongjoon-hyun commented:
I finished my second round. Could you update once more?

SparkQA commented:
**[Test build #87529 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87529/testReport)** for PR 20057 at commit [`b79becd`](https://github.com/apache/spark/commit/b79becd188f67202dd237dddd96456c512e544cd).

SparkQA commented:
**[Test build #87529 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87529/testReport)** for PR 20057 at commit [`b79becd`](https://github.com/apache/spark/commit/b79becd188f67202dd237dddd96456c512e544cd).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/87529/
Test FAILed.

danielvdende commented:
Thanks again @dongjoon-hyun, much appreciated 👍. I've made changes accordingly, and added comments in this PR where appropriate.

SparkQA commented:
**[Test build #87530 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87530/testReport)** for PR 20057 at commit [`7e09f90`](https://github.com/apache/spark/commit/7e09f90ab5625d95d060949a371f03afb50b15cb).

SparkQA commented:
**[Test build #87530 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87530/testReport)** for PR 20057 at commit [`7e09f90`](https://github.com/apache/spark/commit/7e09f90ab5625d95d060949a371f03afb50b15cb).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/87530/
Test FAILed.

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala:
```diff
@@ -85,15 +85,24 @@ private object PostgresDialect extends JdbcDialect {
+      table: String,
+      cascade: Option[Boolean] = isCascadingTruncateTable): String = {
+    cascade match {
+      case Some(true) => s"TRUNCATE TABLE ONLY $table CASCADE"
```
I believe you already these this in PostgreSQL.
To make it sure for us about using these `ONLY` and `CASCADE` at the same time, could you share us the result of a manual test of this with PostgreSQL tables, descendant, and foreign-key referenced tables here?

danielvdende replied:
Sure, I made a quick example, as you can see using `TRUNCATE TABLE ONLY $table CASCADE` will truncate the foreign key-ed table, but leave the inheritance relationship intact:
```postgres=# CREATE SCHEMA test;
CREATE SCHEMA
postgres=# CREATE TABLE parent(a INT);
CREATE TABLE
postgres=# ALTER TABLE parent ADD CONSTRAINT some_constraint PRIMARY KEY(a);
ALTER TABLE
postgres=# CREATE TABLE child(b INT) INHERITS (parent);
CREATE TABLE
postgres=# CREATE TABLE forkey(c INT);
CREATE TABLE
postgres=# ALTER TABLE forkey ADD FOREIGN KEY(c) REFERENCES parent(a);
ALTER TABLE

postgres=# INSERT INTO parent VALUES(1);
INSERT 0 1
postgres=# select * from parent;
 a
---
 1
(1 row)

postgres=# select * from child;
 a | b
---+---
(0 rows)

postgres=# INSERT INTO child VALUES(2);
INSERT 0 1
postgres=# select * from child;
 a | b
---+---
 2 |
(1 row)

postgres=# select * from parent;
 a
---
 1
 2
(2 rows)

postgres=# INSERT INTO forkey VALUES(1);
INSERT 0 1
postgres=# select * from forkey;
 c
---
 1
(1 row)

postgres=# TRUNCATE TABLE ONLY parent CASCADE;
NOTICE:  truncate cascades to table "forkey"
TRUNCATE TABLE
postgres=# select * from parent;
 a
---
 2
(1 row)

postgres=# select * from child;
 a | b
---+---
 2 |
(1 row)

postgres=# select * from forkey;
 c
---
(0 rows)
```

dongjoon-hyun replied:
Yep. Thank you for testing. Could you add more description about this into the parameter description of `cascade`?
```
   * @param cascade Whether or not to cascade the truncation. Default value is the value of
   *                isCascadingTruncateTable()
```

danielvdende replied:
Done :-)

SparkQA commented:
**[Test build #87531 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87531/testReport)** for PR 20057 at commit [`0e646b2`](https://github.com/apache/spark/commit/0e646b20f0be1580cfb8565908c98b57ec248941).

SparkQA commented:
**[Test build #87531 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87531/testReport)** for PR 20057 at commit [`0e646b2`](https://github.com/apache/spark/commit/0e646b20f0be1580cfb8565908c98b57ec248941).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/87531/
Test PASSed.

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/TeradataDialect.scala:
```diff
@@ -31,4 +31,19 @@ private case object TeradataDialect extends JdbcDialect {
+  override def getTruncateQuery(
+      table: String,
+      cascade: Option[Boolean] = isCascadingTruncateTable): String = {
+    s"TRUNCATE TABLE $table"
```
At the beginning, I saw `teradata` is missed in your test case. So, I asked to add that.

But, today, I did double-check that in order to review this line. Unfortunately, according to Teradata document, `TRUNCATE` syntax is not documented. And it's not supported according to some old community articles.

- Teradata SQL Reference
  - https://info.teradata.com/HTMLPubs/DB_TTU_16_00/index.html#page/SQL_Reference%2FB035-1146-160K%2Fpds1472240516459.html%23
- Community (DELETE/TRUNCATE COMMAND)
  - https://community.teradata.com/t5/Database/Delete-truncate-command/td-p/434

dongjoon-hyun replied:
Currently, `TeradataDialect` seems to generate a unsupported syntax because it inherits from `JdbcDialects`. Before doing this PR, we had better check this.

Ping, @klinvill and @gatorsmile who is the original author and committer of this code. Could you give us some comment for this?

dongjoon-hyun replied:
Gentle ping, @klinvill .

klinvill replied:
Hi @dongjoon-hyun, I was the original author of the TeradataDialect and @gatorsmile reviewed and committed it. You are correct, Teradata does not support the TRUNCATE statement. Instead Teradata uses a DELETE statement so you should be able to use `DELETE FROM $table ALL` instead of `TRUNCATE TABLE $table`

dongjoon-hyun replied:
Thank you so much for confirming, @klinvill !

@danielvdende . Could you update `TeradataDialect` according to @klinvill 's advice?

danielvdende replied:
Thanks @klinvill for helping out :-)
@dongjoon-hyun I've made the changes to `TeradataDialect` and updated the tests accordingly.

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/DB2Dialect.scala:
```diff
@@ -49,4 +49,17 @@ private object DB2Dialect extends JdbcDialect {
+      table: String,
+      cascade: Option[Boolean] = isCascadingTruncateTable): String = {
+    s"TRUNCATE TABLE $table"
+  }
```
Since this is the same code from JdbcDialect, it seems we don't need to add this here.
Not only this DB2Dialect, but also DerbyDialect/MsSQLDialect/MySQLDialect.

SparkQA commented:
**[Test build #87543 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87543/testReport)** for PR 20057 at commit [`c262993`](https://github.com/apache/spark/commit/c262993d6cd7bb2a10069d4701693850dd4c9389).

SparkQA commented:
**[Test build #87543 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87543/testReport)** for PR 20057 at commit [`c262993`](https://github.com/apache/spark/commit/c262993d6cd7bb2a10069d4701693850dd4c9389).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/87543/
Test FAILed.

danielvdende committed:
Use correct truncation syntax for Teradata

Also update the tests accordingly

SparkQA commented:
**[Test build #87582 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87582/testReport)** for PR 20057 at commit [`2da0ee7`](https://github.com/apache/spark/commit/2da0ee7293edcc61e4f2e54be7f15368d737b3b3).

SparkQA commented:
**[Test build #87582 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87582/testReport)** for PR 20057 at commit [`2da0ee7`](https://github.com/apache/spark/commit/2da0ee7293edcc61e4f2e54be7f15368d737b3b3).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/87582/
Test PASSed.

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/TeradataDialect.scala:
```diff
@@ -31,4 +31,19 @@ private case object TeradataDialect extends JdbcDialect {
+  override def isCascadingTruncateTable(): Option[Boolean] = Some(false)
+
+  /**
+   * The SQL query used to truncate a table.
```
Please add more description about Teradata's lack of `TRUNCATE` syntax here. Otherwise, people try to make a PR for this mistakenly.

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/TeradataDialect.scala:
```diff
@@ -31,4 +31,19 @@ private case object TeradataDialect extends JdbcDialect {
     case _ => None
   }
+
+  override def isCascadingTruncateTable(): Option[Boolean] = Some(false)
```
Please add a function description that Teradata doesn't support cascading DELETE.
You may need to explain the lack of TRUNCATE syntax, too.

dongjoon-hyun commented on sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala:
```diff
@@ -684,17 +684,17 @@ class JDBCSuite extends SparkFunSuite
   }
 
   test("quote column names by jdbc dialect") {
```
I understand the indention is consistency, but let's keep the unrelated code untouched in this PR.
We can do refactoring PR later if needed.

dongjoon-hyun commented on sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala:
```diff
@@ -805,13 +805,13 @@ class JDBCSuite extends SparkFunSuite
   }
 
   test("PostgresDialect type mapping") {
```
This test case, too.

dongjoon-hyun commented on sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala:
```diff
@@ -839,35 +839,68 @@ class JDBCSuite extends SparkFunSuite
   }
 
   test("table exists query by jdbc dialect") {
```
this test case, too.

dongjoon-hyun commented on sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala:
```diff
@@ -839,35 +839,68 @@ class JDBCSuite extends SparkFunSuite
     assert(derby.getTableExistsQuery(table) == defaultQuery)
   }
 
   test("truncate table query by jdbc dialect") {
```
For this test case, we added `oracle` and `teradata`. It's okay to update this.

dongjoon-hyun commented on sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala:
```diff
@@ -839,35 +839,68 @@ class JDBCSuite extends SparkFunSuite
+    assert(h2.getTruncateQuery(table, Some(true)) == defaultQuery)
+    assert(derby.getTruncateQuery(table, Some(true)) == defaultQuery)
+    assert(oracle.getTruncateQuery(table, Some(true)) == oracleQuery)
+    assert(teradata.getTruncateQuery(table, Some(true)) == teradataQuery)
```
Let's group the same one like the following.
```
Seq(mysql, db2, h2, derby).foreach { dialect =>
    assert(dialect.getTruncateQuery(table, Some(true)) == defaultQuery)
}
```

danielvdende commented:
@dongjoon-hyun Made the changes you pointed out, thanks! 👍

SparkQA commented:
**[Test build #87591 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87591/testReport)** for PR 20057 at commit [`ed452e7`](https://github.com/apache/spark/commit/ed452e755be61d9e0b130e75215e7ab30ceaaed9).

SparkQA commented:
**[Test build #87591 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87591/testReport)** for PR 20057 at commit [`ed452e7`](https://github.com/apache/spark/commit/ed452e755be61d9e0b130e75215e7ab30ceaaed9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/87591/
Test PASSed.

danielvdende commented:
@dongjoon-hyun I saw that Spark 2.3 was released a few days ago, congrats on the release! :-) Is there anything stopping us from merging this PR into master now?

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala:
```diff
@@ -120,14 +121,30 @@ abstract class JdbcDialect extends Serializable {
     s"TRUNCATE TABLE $table"
   }
 
+
```
Let's remove this redundant empty line addition.

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/OracleDialect.scala:
```diff
@@ -95,4 +95,21 @@ private case object OracleDialect extends JdbcDialect {
+      case _ => s"TRUNCATE TABLE $table"
+    }
+  }
+
```
Let's remove line 114.

dongjoon-hyun reviewed (commented):
@gatorsmile . Could you review this PR?

Spark has been supporting `TRUNCATE` before adding `TeradataDialect`. In this PR, it turns out `TeradataDialect` doesn't support `TRUNCATE` SQL syntax at all. Although this PR introduces `DELETE` statement in `TeradataDialect.getTruncateQuery` function, I think this feature is helpful in general.

SparkQA commented:
**[Test build #88040 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88040/testReport)** for PR 20057 at commit [`7e0ff07`](https://github.com/apache/spark/commit/7e0ff071bc366b75b1f77e66a7b7e24a6d5790c9).

SparkQA commented:
**[Test build #88040 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88040/testReport)** for PR 20057 at commit [`7e0ff07`](https://github.com/apache/spark/commit/7e0ff071bc366b75b1f77e66a7b7e24a6d5790c9).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/88040/
Test FAILed.

dongjoon-hyun commented:
Retest this please.

SparkQA commented:
**[Test build #88052 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88052/testReport)** for PR 20057 at commit [`7e0ff07`](https://github.com/apache/spark/commit/7e0ff071bc366b75b1f77e66a7b7e24a6d5790c9).

SparkQA commented:
**[Test build #88052 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88052/testReport)** for PR 20057 at commit [`7e0ff07`](https://github.com/apache/spark/commit/7e0ff071bc366b75b1f77e66a7b7e24a6d5790c9).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/88052/
Test FAILed.

dongjoon-hyun commented:
The failures are
- Last time, `RateSourceV2Suite`.
- This time, `HiveClientSuites`.

dongjoon-hyun commented:
Retest this please.

SparkQA commented:
**[Test build #88057 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88057/testReport)** for PR 20057 at commit [`7e0ff07`](https://github.com/apache/spark/commit/7e0ff071bc366b75b1f77e66a7b7e24a6d5790c9).

SparkQA commented:
**[Test build #88057 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88057/testReport)** for PR 20057 at commit [`7e0ff07`](https://github.com/apache/spark/commit/7e0ff071bc366b75b1f77e66a7b7e24a6d5790c9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

danielvdende commented:
Gentle ping @gatorsmile :-)

danielvdende commented:
Gentle ping again @gatorsmile @dongjoon-hyun :-)

Fokko commented:
Gentle ping @gatorsmile @dongjoon-hyun 👍

danielvdende commented:
@gatorsmile @dongjoon-hyun Guys, any update?

danielvdende committed:
Merge branch 'master' into SPARK-22880

danielvdende commented:
@dongjoon-hyun @gatorsmile Any update, guys?

SparkQA commented:
**[Test build #92841 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92841/testReport)** for PR 20057 at commit [`bc75051`](https://github.com/apache/spark/commit/bc75051f5f4a47ef045c93bd933b2f95635100ad).

SparkQA commented:
**[Test build #92841 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92841/testReport)** for PR 20057 at commit [`bc75051`](https://github.com/apache/spark/commit/bc75051f5f4a47ef045c93bd933b2f95635100ad).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/92841/
Test FAILed.

gatorsmile commented:
retest this please

SparkQA commented:
**[Test build #92963 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92963/testReport)** for PR 20057 at commit [`bc75051`](https://github.com/apache/spark/commit/bc75051f5f4a47ef045c93bd933b2f95635100ad).

SparkQA commented:
**[Test build #92963 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92963/testReport)** for PR 20057 at commit [`bc75051`](https://github.com/apache/spark/commit/bc75051f5f4a47ef045c93bd933b2f95635100ad).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/92963/
Test FAILed.

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala:
```diff
@@ -120,11 +121,26 @@ abstract class JdbcDialect extends Serializable {
+  @Since("2.4.0")
+  def getTruncateQuery(
+    table: String,
+    cascade: Option[Boolean] = isCascadingTruncateTable): String = {
```
Nit: indent.

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala:
```diff
@@ -105,7 +105,12 @@ object JdbcUtils extends Logging {
+          options.isCascadeTruncate))
+      } else {
+        statement.executeUpdate(dialect.getTruncateQuery(options.table))
+      }
```
+1 to the above comment of @dongjoon-hyun

gatorsmile reviewed (commented):
LGTM except two minor comments

gatorsmile commented:
retest this please

SparkQA commented:
**[Test build #93322 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93322/testReport)** for PR 20057 at commit [`bc75051`](https://github.com/apache/spark/commit/bc75051f5f4a47ef045c93bd933b2f95635100ad).

SparkQA commented:
**[Test build #93322 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93322/testReport)** for PR 20057 at commit [`bc75051`](https://github.com/apache/spark/commit/bc75051f5f4a47ef045c93bd933b2f95635100ad).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/93322/
Test FAILed.

danielvdende committed:
Minor fixes

danielvdende commented:
Just made the changes you mentioned @gatorsmile :-)

SparkQA commented:
**[Test build #93327 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93327/testReport)** for PR 20057 at commit [`a365f79`](https://github.com/apache/spark/commit/a365f79b2f29326621a4cd0177780e66c56eaceb).

dongjoon-hyun reviewed (approved):
Thank you, @gatorsmile and @danielvdende .
LGTM, too.

gatorsmile commented:
Thanks! Merged to master.
